### PR TITLE
[recnet-api] Get Users API

### PR DESCRIPTION
**Pull request by joannechen1223**
## Description

Implemented Get Users API:
1. Implement `GET /users`
2. Add logger middleware and log-level control
3. Add a universal exception filter
4. Add Zod schema parser

## Related Issue

https://github.com/lil-lab/recnet/issues/107

## Notes

I'm now using the API models from `@recnet/recnet-api-model`. However, the swagger won't be up to date if we use the provided types without adding the @ApiProperty on each dto parameter.

## Screenshot
TODO

## TODO

- [ ] Investigate swagger solution
- [x] keyword search
- [ ] Unit test

### CODE REVIEW

File: apps/recnet-api/src/utils/error/recnet.error.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  private status: number;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private extra: any | null = null;
+
+  constructor(
+    errorCode: number = ErrorCode.INTERNAL_SERVER_ERROR,
+    status: number = HttpStatus.INTERNAL_SERVER_ERROR,
+    message = "",
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    extra?: any
```
Comment (swh00tw) [inline apps/recnet-api/src/utils/error/recnet.error.ts:16]: What's the use of `extra` exactly?

**Review by swh00tw — APPROVED** [review]

File: apps/recnet-api/src/database/repository/user.repository.ts [hunk 2]
```diff
@@ -1,15 +1,67 @@
...
+    const where: Prisma.UserWhereInput = {};
+    if (filter.handle) {
+      where.handle = filter.handle;
+    }
+    if (filter.id) {
+      where.id = filter.id;
+    }
+
+    if (filter.keyword) {
+      const keywordCondition: Prisma.StringFilter<"User"> = {
```
[the remark this answers is not in the document]
Comment (joannechen1223) [reply apps/recnet-api/src/database/repository/user.repository.ts]: Fixed.
By the way, I found that it doesn't support cross fields search, which was different from what we discussed offline.

File: apps/recnet-api/src/utils/index.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export const getOffset = (page: number, pageSize: number): number =>
+  (page - 1) * pageSize;
```
[the remark this answers is not in the document]
Comment (joannechen1223) [reply apps/recnet-api/src/utils/index.ts:2]: yes, the page is 1-indexed. mentioned in API docs

File: apps/recnet-api/src/utils/error/recnet.error.ts (hunk 1 above)
Comment (joannechen1223) [reply apps/recnet-api/src/utils/error/recnet.error.ts:16]: It provides additional information to the api consumers. For example, I wrapped the zod error message into the extra field.

File: apps/recnet-api/src/utils/filters/recnet.exception.filter.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  HttpStatus,
+  Logger,
+} from "@nestjs/common";
+import { Response } from "express";
+
+import { RecnetError } from "@recnet-api/utils/error/recnet.error";
+import {
+  ErrorCode,
+  errorMessages,
+} from "@recnet-api/utils//error/recnet.error.const";
```
Comment (swh00tw) [inline apps/recnet-api/src/utils/filters/recnet.exception.filter.ts:15]:
```suggestion
} from "@recnet-api/utils/error/recnet.error.const";
```
Typo?

Comment (joannechen1223) [reply apps/recnet-api/src/utils/filters/recnet.exception.filter.ts]: Fixed

**Pull request merged**